### PR TITLE
[Backport 29.0] Bug 648535: Reprice subcontracting lines after scheduling (#10917)

### PR DESCRIPTION
**Pull request by ChethanT**
## Backport of #10917 to `releases/29.0`

Backports [#10917](https://github.com/microsoft/BCApps/pull/10917) — *Bug 648535: Reprice subcontracting lines after scheduling* — to the `releases/29.0` release branch.

### What & why
Subcontracting purchase lines were priced before backward scheduling had finalized the purchase-line Order Date and before routing context was transferred from the requisition line, so a subcontractor price valid for the purchase-header date could be selected instead of the price valid for the final purchase-line date.

The review follow-up also:
- preserves the worksheet's calculated Units/Time cost when no subcontractor price matches;
- preserves manually overridden worksheet costs during carry-out;
- keeps date scheduling editable on released purchase orders without changing financial terms;
- reprices lead-time-only changes when the resulting Order Date crosses a price boundary.

### Source
- Source PR: #10917
- Source squash-merge commit: `049258b463302e36649979d241a6f122c7eb87c0`
- Original backport commit: `c1924162a`
- Corrective follow-up commit: `327c1d82c5`
- Main follow-up PR: #11292
- ADO: [AB#648535](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/648535)

### Cherry-pick / conflict resolution
The product changes applied cleanly. The test file conflicted because the release branch predates the `Subc. Management Library` helper refactor. Only the relevant regression tests were applied, using equivalent local helper procedures already required by this backport. No unrelated `main` tests were introduced. The resolution is identical to the `releases/29.x` backport.

### Validation
- Static verification: no conflict markers or duplicate procedures; `git diff --check` passes.
- Four corrective regressions cover no-price fallback, manual worksheet overrides, released-order date edits, and lead-time-only Order Date changes.
- Full AL compile and test execution are deferred to the AL-Go PR build because local symbol packages are unavailable and sandbox policy blocks CoreXT package-state initialization.

⚠️ Do not merge until the AL-Go PR build passes. **Not** auto-merged.

### CODE REVIEW

**Review by alexei-dobriansky — DISMISSED** [review]

**Comment by ChethanT** [comment]
Applied the same corrective follow-up as #11197 in 327c1d82c5. It preserves calculated no-price fallback costs and manual worksheet overrides, keeps released-order scheduling editable without changing financial terms, and handles lead-time-only Order Date shifts. The four corresponding regression tests are included.

**Comment by alexei-dobriansky** [comment]
## Good Sense Reviewer - Round 1

# **Recommendation: Request Changes**

<details>
<summary><strong>What this PR does</strong></summary>

The change backports the subcontracting repricing fix to the 29.0 branch. It reprices subcontracting purchase lines after requisition fields are transferred and after schedule-related date changes, so the final purchase-line Order Date drives date-effective subcontractor pricing. It also tries to preserve worksheet and manual costs and keeps released-order date edits from changing financial terms.

The BaseApp event publisher for `OnInsertPurchOrderLineOnAfterTransferFromReqLineToPurchLine` exists in the target branch and fires after `TransferFromReqLineToPurchLine` but before the purchase line is inserted, so the new subscriber has the manufacturing fields it needs. The `Purchase Line` date flow assigns `Order Date` directly while validating `Planned Receipt Date`, so the separate Planned Receipt Date and Order Date subscribers are justified. The remaining issue is the new fallback used when no price list entry matches.
</details>

<details>
<summary><strong>Problem-solution fit</strong></summary>

**Fit:** Partial

The date-based pricing issue is covered for creation and rescheduling, and the tests exercise those paths. The no-price fallback is part of the correction, but one new formula is not safe when expected output quantity is 0, so the solution is not complete yet.
</details>

<details>
<summary><strong>Suggestions</strong></summary>

**S1** (🔴 High): *Guard zero expected output before dividing*
When no price matches and the routing cost is time-based, this code divides by `ProdOrderLine."Total Exp. Oper. Output (Qty.)"` with no zero check. A zero expected output quantity will stop subcontracting repricing with a division-by-zero error. Please handle 0 before the division, for example by keeping the current cost or returning a safe fallback.
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** Moderate to high because `Direct Unit Cost` affects subcontracting purchase costs and the new handlers run during purchase-line creation and date validation. The change is scoped to Subcontracting and guarded by feature and open-header checks, but the unguarded no-price fallback can turn a scheduling or carry-out flow into a runtime error. The target-branch BaseApp publisher and purchase-line date flow were verified; the current PR build has not provided a successful validation pass.

**Necessity:** Justified. Without the change, a subcontracting purchase line can keep a price that is valid for a different date than its final Order Date. The backport scope matches the bug, but the fallback path should be made safe before merge.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=4 system=github pr=11198 round=1 by=alexei-dobriansky at=2026-09-10T00:18:57.9626275Z lastSha=327c1d82c581a4dd8cbb00bb27db37b21a5eafa0 reviewKey=90fbd8771802006d8815ef36608ca189d223215b0a9322c5a74ab877dd509949 suggestions=S1@006633ee

**Comment by ChethanT** [comment]
Applied the accepted #11292 review follow-up in `eae59488df`: zero-output fallback guard, pre-lookup applicability guards, partial routing-line loading, and direct no-price fallback coverage for Units, Time, and zero expected output. The suggested two-`Get` routing lookup was intentionally not adopted because the existing `Released|Finished` `FindFirst` is one bounded clustered-key query.

File: src/Apps/W1/Subcontracting/App/src/Purchase/SubcPriceManagement.Codeunit.al [hunk 1]
```diff
@@ -521,36 +519,89 @@ codeunit 20508 "Subc. Price Management"
...
+            exit(false);
 
-        PurchaseLine."Direct Unit Cost" := DirectCost;
-        PurchaseLine.Validate("Line Discount %");
+        if SubcontractorPrice."Unit of Measure Code" = PurchaseLine."Unit of Measure Code" then
+            PriceListUOM := SubcontractorPrice."Unit of Measure Code";
+        GetUOMPrice(PurchaseLine."No.", GetQuantityBase(PurchaseLine), SubcontractorPrice, PriceListUOM, PriceListQtyPerUOM, PriceListQty);
+        GetPriceByUOM(SubcontractorPrice, PriceListQty, PriceListCost);
+        if PriceListCost = 0 then
+            exit(true);
```
[the remark this answers is not in the document]
Comment (ChethanT) [reply src/Apps/W1/Subcontracting/App/src/Purchase/SubcPriceManagement.Codeunit.al:587]: Addressed in `12de38cc87`. `GetPriceByUOM` now distinguishes no applicable minimum-quantity tier from a matched zero-cost tier. The former uses the standard subcontracting fallback; the latter remains a valid zero price. Added regression coverage for both cases.

**Comment by alexei-dobriansky** [comment]
## Good Sense Reviewer - Round 2

# **Recommendation: Request Changes**

<details>
<summary><strong>What this PR does</strong></summary>

The new commits address the previous zero-output fallback issue and add coverage for no-price fallbacks, minimum-quantity tiers, released-line date edits, and lead-time-only rescheduling. The original date-based repricing path is still mostly in place: the creation subscriber compares the transferred cost with the requisition-date price before repricing with the final purchase-line Order Date, and the date subscribers reprice open subcontracting lines when Order Date changes.

The zero-output division is now guarded, but the new minimum-quantity false return leaves a carry-out path where final-date repricing can be skipped. If the requisition-date price record has no applicable minimum-quantity tier, the creation subscriber exits before checking the final purchase-line Order Date.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|----|-------|--------|-----------------|
| S1 | Guard zero expected output before dividing | Addressed | A zero-output guard and regression test were added. |
</details>

<details>
<summary><strong>New observations (commits since round 1)</strong></summary>

**S2** (🔴 High): *Carry-out can skip final-date pricing after a quantity miss*
The new false return from `TryGetSubcPriceListCostForPurchLine` makes the creation subscriber exit when the requisition-date price has no applicable minimum-quantity tier. That happens before `GetSubcPriceForPurchLine` can reprice with the final purchase-line `Order Date`. In that case, a valid final-date price can be missed and the purchase line can keep 0 or another worksheet cost.
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** High because the affected value is `Direct Unit Cost` on subcontracting purchase lines, and the missed branch can persist a wrong purchase cost. The change is still scoped to Subcontracting and the target-branch publisher and date flow were verified, but the minimum-quantity/date-effective combination remains inside the main carry-out path. The remote build was still running when this review was rendered; `git diff --check` passed locally.

**Necessity:** The change is justified because subcontracting pricing must use the final purchase-line Order Date after backward scheduling. The scope is right for the reported bug, but the carry-out guard should still allow final-date repricing when the requisition-date lookup fails only because no minimum-quantity tier applies.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=4 system=github pr=11198 round=2 by=alexei-dobriansky at=2026-09-10T12:32:05.6985192Z lastSha=12de38cc875495f0c04220bb392ed09b6e42ee75 reviewKey=077396b0c86fd8e545011da5a2193931d8d10a866368479cd531cd0d5d7e456d suggestions=S1@006633ee:addressed,S2@d13a1cdd:new parentRound=1

**Comment by ChethanT** [comment]
Synchronized the latest #11292 review fixes in `88998c83f0`. This includes requisition-context automatic-cost detection, final purchase-line date pricing, minimum-quantity fallback handling, strict missing-routing diagnostics, and release-compatible regression tests.

**Comment by alexei-dobriansky** [comment]
## Good Sense Reviewer - Round 3

# **Recommendation: Request Changes**

<details>
<summary><strong>What this PR does</strong></summary>

The latest commit addresses the previous carry-out pricing gap by comparing the requisition line with the automatic subcontracting cost, then repricing the purchase line with the final line Order Date when the cost was not manually overridden. It also keeps the fallback path for no applicable price and adds more regression coverage for final-date pricing and minimum-quantity fallback.

The pricing flow now matches the reported scenario: automatic worksheet costs can be replaced by the price valid on the final purchase-line date, while manual worksheet overrides are preserved. The remaining blocker is not the pricing logic itself; the new test code does not compile on the target branch because a library variable was not backported.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|----|-------|--------|-----------------|
| S1 | Guard zero expected output before dividing | Addressed | The zero-output guard and regression test are still present. |
| S2 | Carry-out can skip final-date pricing after a quantity miss | Addressed | The creation path now uses automatic-cost detection and still calls final-date purchase-line pricing when the requisition-date lookup has no applicable tier. |
</details>

<details>
<summary><strong>New observations (commits since round 2)</strong></summary>

**S3** (🔴 High): *Declare LibraryRandom for new tests*
The new tests use `LibraryRandom`, but this test codeunit does not declare it on the target branch. The W1 compile fails with AL0118 for this symbol. Add `LibraryRandom: Codeunit "Library - Random";` to the global var block, or replace the new random calls.
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** High until the compile error is fixed, because the Subcontracting test app cannot build. The functional change still touches `Direct Unit Cost` on subcontracting purchase lines, so the added regressions need to compile and run before merge.

**Necessity:** The change is justified because subcontracting pricing must use the final purchase-line Order Date after scheduling. The latest product code is targeted and matches the needed behavior, but the backport is not merge-ready while its tests reference an undeclared library.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=4 system=github pr=11198 round=3 by=alexei-dobriansky at=2026-09-11T09:35:48Z lastSha=88998c83f040e805b6bed627f8407d96bf25fff6 reviewKey=880c570da34275300e5c86317a7b9a9a089456204926e03a8954e2b6f6250227 suggestions=S1@006633ee:addressed,S2@d13a1cdd:addressed,S3@0f191555:new parentRound=2

**Comment by alexei-dobriansky** [comment]
## Good Sense Reviewer - Round 4

# **Recommendation: Accept with Suggestions**

<details>
<summary><strong>What this PR does</strong></summary>

Since round 3, the change declares the missing `LibraryRandom` test-library variable and guards minimum-amount pricing so zero quantity no longer divides by `PriceListQty`. The earlier subcontracting repricing flow still matches the reported bug: automatically calculated worksheet costs can be replaced by the price valid on the final purchase-line Order Date, while manual worksheet costs and released-order financial terms are preserved.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|----|-------|--------|-----------------|
| S1 | Guard zero expected output before dividing | Addressed | The zero-output guard and regression test remain present. |
| S2 | Carry-out can skip final-date pricing after a quantity miss | Addressed | The creation path now compares against the automatic requisition-line cost and still reprices with the final purchase-line Order Date when the requisition-date tier does not apply. |
| S3 | Declare LibraryRandom for new tests | Addressed | `LibraryRandom: Codeunit "Library - Random";` is now declared in the global var block. |
</details>

<details>
<summary><strong>New observations (commits since round 3)</strong></summary>

**S4** (🟠 Moderate): *Cover zero-quantity minimum amount pricing*
The new guard prevents dividing by zero when `PriceListQty` is 0, but the added tests do not exercise a matching price with `Minimum Amount` set and zero quantity. Add a small regression test for that branch so future changes do not reintroduce the runtime error.
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** Moderate because the changed code still sets `Direct Unit Cost` on subcontracting purchase lines, and the newest change is in minimum-amount price calculation. The direct divide-by-zero risk is fixed, prior blockers are addressed, and live checks showed no completed failing status checks, but many build/test jobs were still running when this review was prepared.

**Necessity:** Justified. Subcontracting pricing must use the final purchase-line Order Date after scheduling, and the round-4 changes are targeted cleanup for review-found issues.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=4 system=github pr=11198 round=4 by=alexei-dobriansky at=2026-09-16T12:06:01Z lastSha=2495aecc15c866b352f3a4710dc723651882e29d reviewKey=ef5febf4216215558edf3f9b2b64b6cae9b4b3d50ada0d1e6f325abc85935e7b suggestions=S1@006633ee:addressed,S2@d13a1cdd:addressed,S3@0f191555:addressed,S4@ce42eda0:new parentRound=3

**Comment by alexei-dobriansky** [comment]
## Good Sense Reviewer - Round 5

# **Recommendation: Accept**

<details>
<summary><strong>What this PR does</strong></summary>

Since round 4, the change adds the missing regression test for zero quantity with `Minimum Amount`. The test creates a matching subcontractor price, sets purchase quantity to zero, calls the repricing path, and checks that the price list direct unit cost is kept. The full change still matches the reported bug: subcontracting purchase lines are repriced after scheduling sets the final Order Date, while no-price fallback, manual worksheet cost overrides, and released-order schedule edits are preserved.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|----|-------|--------|-----------------|
| S1 | Guard zero expected output before dividing | Addressed | The zero-output guard and regression test remain present. |
| S2 | Carry-out can skip final-date pricing after a quantity miss | Addressed | The creation path still reprices from the final purchase-line date when the automatic worksheet cost was used. |
| S3 | Declare LibraryRandom for new tests | Addressed | The test-library variable remains declared. |
| S4 | Cover zero-quantity minimum amount pricing | Addressed | The new test covers a matching price with `Minimum Amount` and zero purchase quantity. |
</details>

<details>
<summary><strong>New observations (commits since round 4)</strong></summary>

*None - changes only addressed prior suggestions.*
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** Moderate because the full change updates `Direct Unit Cost` on subcontracting purchase lines and runs from purchase-line date validation and requisition carry-out. The new round is test-only. The required publisher exists on the target branch, it fires after requisition values are copied to the purchase line, and the date-validation flow has already recalculated Order Date before the subscriber runs.

**Necessity:** Justified. Without this change, a subcontracting purchase line can keep a subcontractor price that was selected for a different effective date than the final purchase-line Order Date. Users would need to notice and correct the cost manually.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=4 system=github pr=11198 round=5 by=alexei-dobriansky at=2026-09-16T18:17:50Z lastSha=24c487155290180a95e57ba47ad994652de3a5c8 reviewKey=40c973f8e178c53b9f5afd0c2caedbfe2a494ec9570e2d68759e0da93fa5df03 suggestions=S1@006633ee:addressed,S2@d13a1cdd:addressed,S3@0f191555:addressed,S4@ce42eda0:addressed parentRound=4

**Review by alexei-dobriansky — APPROVED** [review]

**Review by attilatoury — APPROVED** [review]
